### DISCUSSION TITLE
Battle for Wesnoth port

### DISCUSSION THREAD

**Discussion by oherau** [OnionUI/Onion#1531] [Ideas]
Hi there

Battle for Wesnoth is one of the best open source games I know.
It's a turn based RPG taking place in a kinda fantasy/medieval age.
The whole game is played on an hexagonal grid, like "Heros of Might & Magic series" but without close range battles.
It's quite old (2003), but updated regularly since then.
The code is GPL 2.0, the game is multiplateforms with nice graphics, tons of features, an online mode and a solo campain.

It's a solid game which worth to be played, and would be a nice addition to ports
[https://github.com/wesnoth](https://github.com/wesnoth)

**Comment by Aemiii91** [comment]
I don't see this happening, mainly because:

Minimum requirements: **2 GB of RAM**

The Miyoo Mini and Mini+ only have **128 MB**.

**Reply by oherau** [reply]
ah you are totally right :(

Edit: it came to my mind that other less demanding opensource and multiplateform games might be a nice addition to the port list
[https://en.wikipedia.org/wiki/List_of_open-source_video_games](https://en.wikipedia.org/wiki/List_of_open-source_video_games)

**Comment by michaelachrisco** [comment]
Looks like if you take an older version of Westnoth it might work: 

What are the system requirements?

outdated! We are not completely certain, but an x86 running at 400 MHz with 128 MB RAM should be adequate for versions 1.0.2 and below. For versions 1.1 and up we recommend a computer with at least 1 GHz and 512 MB RAM if you run KDE or Gnome as Windowmanager (The game itself needs about 100 MB RAM). Slower machines will have trouble scrolling large maps or processing AI turns with many units. See the [forum thread](http://www.wesnoth.org/forum/viewtopic.php?t=7384) about minimum and recommended system requirements.

**Comment by michaelachrisco** [comment]
Its very old but looks like it might (big might) work with the older version: https://github.com/wesnoth/wesnoth/releases/tag/1.0.2 

https://github.com/wesnoth/wesnoth/blob/1.0/INSTALL 

Install looks like the custom toolchain might compile? Maybe?

**Comment by michaelachrisco** [comment]
This is as far as I got: 

https://github.com/michaelachrisco/wesnoth/releases/tag/mmp-nonworking

zip contains docker with all the files I had to port in manually is here. Its kinda a mess but its late.  

I was able to get farther than before but hit a snag with a very old libsdl-mixer1.2-dev library. I was able to compile and link every other library, but I cant seem to find that particular library. Usually I would solve this by:
apt install libsdl-mixer1.2-dev
But of course its not available: 
```
root@9b2566286956:~/workspace# apt install libsdl-mixer1.2-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libsdl-mixer1.2-dev
E: Couldn't find any package by glob 'libsdl-mixer1.2-dev'
E: Couldn't find any package by regex 'libsdl-mixer1.2-dev'
```

Error in question: 
```
./configure
...

checking for SDL_mixer.h... no
configure: error: *** SDL_mixer include files not found!
You should install development package.
```


Im trying to find the source so I can do a quick ./configure, make, make install on the dev, but it may be too old? Im stuck.

Install instructions here: https://github.com/wesnoth/wesnoth/blob/1.0/INSTALL

Anyways if anyone else wants to pick this up, feel free to take what Ive built and run with it. Theres probably something simple im not getting.

**Reply by oherau** [reply]
This might help you:
https://askubuntu.com/questions/483302/unable-to-locate-package-libsdl1-2-dev

**Reply by michaelachrisco** [reply]
Thanks, I tried it with the following changes to the Dockerfile: 
```
# Packages
RUN apt-get -y update
RUN apt-get -y install \
	bc \
	build-essential \
	bzip2 \
	bzr \
	cmake \
	cmake-curses-gui \
	cpio \
	git \
	libncurses5-dev \
	make \
	p7zip-full \
	rsync \
	scons \
	sqlite3 \
	tree \
	unzip \
	wget \
	zip \
	locales \
	libsdl1.2-dev \
	libfreetype6-dev \
	libxt-dev \
	libsdl-image1.2-dev \
	libsdl-mixer1.2-dev \
	libsdl-ttf2.0-dev \
	libsdl-gfx1.2-dev \
	libgtest-dev \
	cppcheck
```
Looks like the same error. And im assuming we cant just do that for the port, we will need the source libraries (I think?).

**Reply by oherau** [reply]
When does the error occurs? During sdl 1.2 package install during docker RUN ? Or during the ./configure or make?
I think it would be a better option to stick to the lastest version (if possible with minor adjustments) or give up

**Reply by michaelachrisco** [reply]
With the help of XK and Vitty85 on discord, I stumbled into a solution. Then XK made a build-chain that's much better. 

```
git clone https://github.com/shauninman/union-miyoomini-toolchain.git
cd union-miyoomini-toolchain
make shell
``` 

then copy this script over to the `~/workspace` folder

```bash
unset urls
unset files
unset log_files
unset missing_tools

echo "******************************************"
echo "*                                        *"
echo "*             PREPARING....              *"
echo "*                                        *"
echo "*                                        *"
echo "******************************************"

# pull some packages from the sources

package_installer() {
    dpkg --add-architecture armhf
    apt-get update

    packages=(
      "libsdl1.2-dev:armhf"
      "libsdl-image1.2-dev:armhf"
      "libsdl-mixer1.2-dev:armhf"
      "libsdl-net1.2-dev:armhf"
      "libfreetype2-dev:armhf"
    )

    for pkg in "${packages[@]}"; do
      if ! dpkg -l | grep -q "$pkg"; then
        echo "Installing missing package: $pkg"
        apt-get install -y "$pkg"
      else
        echo "Package $pkg is already installed."
      fi
    done
}

package_installer

# check which dev packages we have and if they work, if not install them
check_dev_tools() {
  declare -A tool_to_package_map
  tool_to_package_map["pkg-config"]="pkg-config"
  tool_to_package_map["autoconf"]="autoconf"
  tool_to_package_map["autoupdate"]="autoconf"
  tool_to_package_map["libtoolize"]="libtool"
  tool_to_package_map["m4"]="m4"
  tool_to_package_map["automake"]="automake"
  tool_to_package_map["autoreconf"]="autoconf"

  missing_tools=()
  missing_packages=()

  for tool in "${!tool_to_package_map[@]}"; do
    if command -v "$tool" > /dev/null 2>&1; then
      echo "$tool exists."
    else
      echo "$tool does not exist."
      package=${tool_to_package_map[$tool]}
      missing_tools+=("$tool")
      if ! [[ " ${missing_packages[@]} " =~ " ${package} " ]]; then
        missing_packages+=("$package")
      fi
    fi
  done

  if [ ${#missing_tools[@]} -ne 0 ]; then
    echo "Missing tools: ${missing_tools[*]}"
    echo "Corresponding missing packages: ${missing_packages[*]}"
    echo "Running 'apt-get update' and then 'apt-get install' for each missing package."
    apt-get update
    apt-get install -y "${missing_packages[@]}"
  else
    echo "All tools are available and functional."
  fi
}

check_dev_tools

# setup cross compile environment - we'll dump it in /mnt for now
export ROOTDIR="${PWD}"
export BIN_NAME="Wesnoth"
export SD_DIR="App"
export FIN_BIN_DIR="/mnt/SDCARD/$SD_DIR/$BIN_NAME"

# setup the CC environment, link correct dir
# set some optimisations
export CROSS_COMPILE="arm-linux-gnueabihf"
export AR=${CROSS_COMPILE}-ar
export AS=${CROSS_COMPILE}-as
export LD=${CROSS_COMPILE}-ld
export RANLIB=${CROSS_COMPILE}-ranlib
export CC=${CROSS_COMPILE}-gcc
export NM=${CROSS_COMPILE}-nm
export HOST=arm-linux-gnueabihf
export BUILD=x86_64-linux-gnu
export CFLAGS="-Wno-undef -Os -marm -mtune=cortex-a7 -mfpu=neon-vfpv4  -march=armv7ve+simd -mfloat-abi=hard -ffunction-sections -fdata-sections -I/opt/miyoomini-toolchain/arm-linux-gnueabihf/libc/usr/include/freetype2/ -I/usr -I/opt/miyoomini-toolchain/arm-linux-gnueabihf/libc/usr/include/"
export CPPFLAGS="-I/opt/miyoomini-toolchain/arm-linux-gnueabihf/libc/usr/include/ -I/opt/miyoomini-toolchain/arm-linux-gnueabihf/include"
export CXXFLAGS="-O3 -fPIC -pthread"
export PATH="$PATH:$FIN_BIN_DIR/bin"
export LDFLAGS="-L/opt/miyoomini-toolchain/arm-linux-gnueabihf/libc/usr/lib/ -L/usr/lib/arm-linux-gnueabihf" # try and use the dockers libSDLS, if not fall back to the ones we pulled from apt
export PKG_CONFIG_PATH="/opt/miyoomini-toolchain/arm-linux-gnueabihf/libc/usr/lib/pkgconfig"
## env setup done, move onto compiling

echo -e "-Starting shortly - a full logfile with be in: \033[32m"$LOGFILE "\033[0m"
echo -e "\n"

cd ~/workspace/

echo "******************************************"
echo "*                                        *"
echo "*             BUILDING Wesnoth           *"
echo "*                                        *"
echo "*                                        *"
echo "******************************************"

cd ~/workspace

# make the output dir
mkdir -p "$FIN_BIN_DIR"

mkdir logs

# Start compiling..
# clone and setup the build for wesnoth, if the dir already exists, skip the clone but run a make clean, then an autogen. 

export REPO_URL="https://github.com/wesnoth/wesnoth.git"
export BRANCH="1.0"
DIR="$ROOTDIR/wesnoth"

if [ ! -d "$DIR" ]; then
    echo -e "\n\n\n Cloning Wesnoth $BRANCH branch\n\n\n"
    git clone $REPO_URL -b $BRANCH $DIR
    cd $DIR
    ./autogen.sh
    wait 
    ./configure CC=$CC CXX=$CXX --host=$HOST --build=$BUILD --prefix=$FIN_BIN_DIR
    wait
else
    echo "\n\n\n Wesnoth directory exists. Preparing for rebuild.\n\n\n"
    cd $DIR
    make clean
    ./autogen.sh
    wait
    ./configure CC=$CC CXX=$CXX --host=$HOST --build=$BUILD --prefix=$FIN_BIN_DIR
    wait
fi

make -j$(( $(nproc) - 2 )) && make install -j$(( $(nproc) - 2 ))

cd "$ROOTDIR"

# check if it actually compiled
if [ -f "$FIN_BIN_DIR/bin/wesnoth" ]; then
    echo -e "\033[32mWesnoth installation successful.\033[0m"
else
    echo -e "\033[31mWesnoth installation failed. Binary not found in $FIN_BIN_DIR/bin.\033[0m"
fi

# if wesnoth installation check is successful, proceed to tar and copy to the root of the workspace
ARCHIVE_NAME="wesnoth_$(date +%Y-%m-%d).tar.gz"
tar -czf "$ARCHIVE_NAME" -C "$FIN_BIN_DIR" .
cp "$ARCHIVE_NAME" $ROOTDIR
echo -e "\033[32m Archive $ARCHIVE_NAME has been copied to /root/workspace.\033[0m"
```

./setup.sh

watch everything get compiled and tar'ed up. 

I did very little of the above, all credit goes to MK. At most I found out what version might work with the MMP.

**Reply by oherau** [reply]
Great job! Anyone installed and tested it?

**Comment by XK9274** [comment]
Vitty, Mike and I have done some work on it but this is becoming a massive task so i'm opening it up here for people to get involved as i'm not sure how much time i can put into it - there's an appdist here that will boot the game to the title screen but mouse support needs adding, the UI needs scaling and probably some other stuff needs fixing. 

**New make script:** https://www.dropbox.com/scl/fi/74ky1wvzt8fef7tg6lbpq/mk.sh?rlkey=apa2nisztflqc9aydmgct6vll&dl=0
**09/04/2024 Dist:** https://www.dropbox.com/scl/fi/idh6rlzzv003iq4y77d02/Wesnoth.7z?rlkey=tiwyxjkaox7ro8d6ax9prc3j2&dl=0
**Repo:** https://github.com/XK9274/wesnoth/tree/1.0

## Instructions:

```
git clone https://github.com/shauninman/union-miyoomini-toolchain.git
cd union-miyoomini-toolchain
make shell
```

Copy mk.sh to ~/workspace

run:

``` 
source mk.sh
```

**Comment by oherau** [comment]
Gamepad support was suggested a very long time ago.
If there is a chance this feature is eventually released then the port would be much easier. Otherwise I think this port can go down in priority and/or left to be completed by volonteers.

Anyway there a tons of other good opensource turn-based strategy games which port can be much easier to do.

**Comment by michaelachrisco** [comment]
Heres a working build: https://github.com/XK9274/wesnoth/releases/tag/v2.0

Big thanks to @Vitty85 and @XK9274

**Reply by oherau** [reply]
Amazing ! Apart from a few graphical glitches it is definitely playable !
Great job guys !
